src/app/application/services/data-source/index.ts

```diff
@@ -1,36 +1,97 @@
-import CallableEventSource, { isClosed } from './CallableEventSource'
+import CallableEventSource from './CallableEventSource'
 import type { Creator, Destroyer } from './DataSourcePool'
+export type { DataSourceResponse } from './DataSourcePool'
+
+type Configuration = {
+  interval?: number
+  retry?: (e: any) => Promise<any> | undefined
```

Comment:
**Issue**: `e` shouldn’t be typed as `any`. We seem to be always passing it a an instance of `ApiError` so why not use that?

Also, it doesn’t look like we ever intent to return data with the retry promise so it should just be using `void`

```suggestion
  retry?: (e: ApiError) => Promise<void> | undefined
```

Also, why can the retry callback return `undefined`?

Reply:
> Issue: e shouldn’t be typed as any. We seem to be always passing it a an instance of ApiError so why not use that?

I think this has been covered in another comment, lmk if not.

> Also, it doesn’t look like we ever intent to return data with the retry promise so it should just be using void

Yeah maybe, maybe not. I'm still not totally sure what this should return longer term.

> Also, why can the retry callback return undefined?

Because if it doesn't return a promise its like saying "in this case, don't retry". Currently `retry` can return `undefined` (not Promise) specifically for this reason.

---

Overall you can probably see from the todo here (and the PR description) I might change the shape of this slightly, but that will become clearer once we are using this more. For the moment, I wanted to just get the feature PRed as I had already spent a bit of time noodling on exactly how `retry` should look, again that will become clearer if/when we use it more.

Is there anything here that is essential to change? I think at a push the only change I'd consider making here is changing `any` to `void` in the Typescript, but even then I'm not too concerned right now.

lmk

Reply:
I changed the `Promise<any>` to `Promise<void>` here, I think we are resolved with this now, but let me know if not!

Reply:
Can we also change `e` to be `unknown` so we’re at least forced to properly access whether it has properties we’re accessing (preferably via an `instanceof` check).

Reply:
I changed `e` to `unknown`, a `instanceof` check won't cut it as we aren't checking for an instance we are checking for existence of a `status` property, it can be an instance of anything as long as it has a `status` property.

FYI: this is likely to change entirely pretty soon.

Resolving this convo, lmk if it needs reopening.